Add configurable UUID conversion for non-AWS SQS-compatible services

jm0514 opened the pull request:
# Add configurable UUID conversion for non-AWS SQS-compatible services

## :loudspeaker: Type of change
<!--- Put an `x` in the boxes that apply -->
- [x] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring

## :scroll: Description
Added configurable support for SQS-compatible cloud services (like Yandex Message Queue) that use non-UUID MessageId formats. The framework now supports both UUID and non-UUID MessageId handling through a new configuration option.

### Configuration:
```properties
# For AWS SQS (default behavior)
spring.cloud.aws.sqs.convert-message-id-to-uuid=true

# For non-AWS SQS-compatible services (Yandex, etc.)
spring.cloud.aws.sqs.convert-message-id-to-uuid=false
```

## :bulb: Motivation and Context
Solves #814 

## :green_heart: How did you test it?
- **Unit Tests**: Added comprehensive tests for `SqsHeaderMapper` and `MessageHeaderUtils`

## :pencil: Checklist
<!--- Put an `x` in the boxes that apply -->
- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [ ] I updated reference documentation to reflect the change
- [x] All tests passing
- [x] No breaking changes

## :crystal_ball: Next steps

jm0514 commented on spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/support/converter/SqsHeaderMapper.java:
```diff
@@ -156,9 +163,27 @@ public MessageHeaders toHeaders(Message source) {
+			accessor.setHeader(SqsHeaders.SQS_AWS_MESSAGE_ID_HEADER, source.messageId());
+			MessageHeaders messageHeaders = accessor.toMessageHeaders();
+			logger.trace("Mapped headers {} for message {}", messageHeaders, source.messageId());
+			return new MessagingMessageHeaders(messageHeaders, UUID.randomUUID());
```
Thanks for the suggestion! Updated to generate UUID from Message ID instead of random UUID.

tomazfernandes commented:
Hey @jm0514, thanks for the PR!

> Perhaps MessagingMessageHeaders can have id set as String instead of UUID? cc @tomazfernandes

@maciejwalkowiak, the issue there is that the `MessagingMessageHeaders` inherits from `MessageHeaders` from Spring, and that one requires an `UUID` type in the constructor.

Might be worth raising with the Spring Messaging team?

In the meantime, I think we can make the solution simpler if we leverage the fact that most ID-related operations are made through the `MessageHeaderUtils#getId` method, which already returns a `String`.

Instead of requiring a configuration change to support other UUID types, we could simplify the logic by trying to parse the UUID, and if it fails we add the new header as suggested in the PR.

Then in the `MessageHeaderUtils#getId` method, we check for the header first, and return either that, or the Spring Message ID if the former is absent.

The caveat would be that for non UUID ids, the result from getting the ID header directly would be a random UUID, but I don't think there's a way around it with the existing `MessageHeaders` implementation, so properly documenting the behavior should suffice. And since we don't currently support that anyway, we wouldn't really need a breaking change.

What do you folks think?

tomazfernandes commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/sqs/SqsAutoConfiguration.java:
```diff
@@ -159,10 +161,15 @@ private void configureProperties(SqsContainerOptionsBuilder options) {
-		public MessagingMessageConverter<Message> messageConverter(ObjectProvider<JsonMapper> jsonMapperProvider) {
-			JsonMapper jsonMapper = jsonMapperProvider.getIfAvailable();
-			return jsonMapper != null ? new SqsMessagingMessageConverter(jsonMapper)
+		public MessagingMessageConverter<Message> messageConverter(ObjectProvider<JsonMapper> jsonMapperProvider,
```
Currently the auto-config manually creates a SqsHeaderMapper and sets the flag on it. Instead, let's add `convertMessageIdToUuid` as a property on both `SqsContainerOptions` and `SqsTemplateOptions`, so  auto-config just does:

```
  factory.configure(options -> options.convertMessageIdToUuid(sqsProperties.getConvertMessageIdToUuid()));
  builder.configure(options -> options.convertMessageIdToUuid(sqsProperties.getConvertMessageIdToUuid()));
```

This keeps the configuration surface consistent between container and template, removes the manual `SqsHeaderMapper` creation from auto-config, and gives programmatic users (without auto-config) a clean way  to set it through the standard options API.

jm0514 replied:
Thanks for the great suggestion! This makes the configuration much cleaner and consistent with the existing patterns.

Applied in 1d32b0c1. Moved convertMessageIdToUuid to both SqsContainerOptions and SqsTemplateOptions, so auto-config now uses factory.configure() and builder.configure() instead of manually creating SqsHeaderMapper. 
Added getHeaderMapper() to AbstractMessagingMessageConverter to propagate the option to the header mapper.

tomazfernandes commented on spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/support/converter/SqsHeaderMapper.java:
```diff
@@ -157,9 +164,36 @@ public MessageHeaders toHeaders(Message source) {
-		MessageHeaders messageHeaders = accessor.toMessageHeaders();
-		logger.trace("Mapped headers {} for message {}", messageHeaders, source.messageId());
-		return new MessagingMessageHeaders(messageHeaders, UUID.fromString(source.messageId()));
+
```
Let's consolidate this logic so we get consistent behavior across both listener and template.

My suggestion would be to create a `SqsMessageIdResolver` utility class with a method such as:

```java
 public static MessageHeaders resolveAndAddMessageId(String messageId, MessageHeaders headers,
          boolean convertMessageIdToUuid) {
      MessageHeaders withRawId = MessageHeaderUtils.addHeaderIfAbsent(
          headers, SqsHeaders.SQS_RAW_MESSAGE_ID_HEADER, messageId);
      if (isValidUuid(messageId)) {
          return new MessagingMessageHeaders(withRawId, UUID.fromString(messageId));
      }
      if (convertMessageIdToUuid) {
          throw new MessagingException(String.format(
              "Message ID '%s' is not a valid UUID. To support non-UUID message IDs, "
                  + "set 'spring.cloud.aws.sqs.convert-message-id-to-uuid=false'. "
                  + "The raw message ID will be available via the '%s' header.",
              messageId, SqsHeaders.SQS_RAW_MESSAGE_ID_HEADER));
      }
      return new MessagingMessageHeaders(withRawId,
          UUID.nameUUIDFromBytes(messageId.getBytes(StandardCharsets.UTF_8)));
  }

 private static boolean isValidUuid(String value) {
          try {
              UUID.fromString(value);
              return true;
          }
          catch (IllegalArgumentException e) {
              return false;
          }
      }
```

Then we'd use it in both `SqsHeaderMapper` and `SqsTemplate`.

Open to suggestions if you have a different idea.

jm0514 replied:
Your suggestion was clear and well thought out, so I followed it as proposed.

Applied in f1d7a50c. Created SqsMessageIdResolver with resolveAndAddMessageId() for the listener side and resolveUuid() / isValidUuid() for the template side. 
Both SqsHeaderMapper and SqsTemplate now delegate to this utility class.

tomazfernandes commented on spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/operations/SqsTemplate.java:
```diff
@@ -327,16 +342,22 @@ private CompletableFuture<Boolean> handleAutoDeduplication(String endpointName)
+		if (sequenceNumber != null) {
+			additionalInfo.put(SqsTemplateParameters.SEQUENCE_NUMBER_PARAMETER_NAME, sequenceNumber);
+		}
+		UUID messageId = SqsMessageIdResolver.resolveUuid(rawMessageId);
```
I'd suggest always adding the new header to `originalMessage` through `resolveAndAddMessageId`, then reading the UUID value back from the resulting message.

That keeps the behavior consistent between receive and send paths and avoids silently handling the UUID conversion when not explicitly configured.

That also allows us to make the remaining methods of `SqsMessageIdResolver` private.

jm0514 replied:
`createSendResult()` now delegates to `resolveAndAddMessageId()`, which adds `SQS_RAW_MESSAGE_ID_HEADER `to the message headers and applies the same fail-fast behavior as the receive path.

tomazfernandes commented on spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/operations/SqsTemplateParameters.java:
```diff
@@ -39,4 +39,9 @@ public class SqsTemplateParameters {
+	/**
+	 * The raw provider message ID when it is not a valid UUID.
+	 */
+	public static final String RAW_MESSAGE_ID_PARAMETER_NAME = "rawMessageId";
```
Let's name it "SQS_RAW_MESSAGE_ID_PARAMETER_NAME" for consistency with the header.

jm0514 replied:
The `RAW_MESSAGE_ID_PARAMETER_NAME `constant was removed since `createSendResult()` now delegates to `resolveAndAddMessageId()`, which stores the raw ID in `MessageHeaders` via `SQS_RAW_MESSAGE_ID_HEADER`. 
This made the `additionalInfo` usage and the constant itself unnecessary.

tomazfernandes commented on spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/operations/SqsTemplate.java:
```diff
@@ -327,16 +342,22 @@ private CompletableFuture<Boolean> handleAutoDeduplication(String endpointName)
+		}
+		UUID messageId = SqsMessageIdResolver.resolveUuid(rawMessageId);
+		if (!SqsMessageIdResolver.isValidUuid(rawMessageId)) {
+			additionalInfo.put(SqsTemplateParameters.RAW_MESSAGE_ID_PARAMETER_NAME, rawMessageId);
```
If this id is already added to the `MessageHeaders`, we may not need to also expose it through `additionalInfo`.

If we keep this parameter, we should always add it so users can rely on it regardless of whether they are using SQS or another implementation.

tomazfernandes commented on spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/support/converter/AbstractMessagingMessageConverter.java:
```diff
@@ -119,6 +120,15 @@ public void setPayloadTypeHeaderValueFunction(Function<Message<?>, String> paylo
+	 * {@link software.amazon.awssdk.services.sqs.model.Message} instances.
+	 * @return the header mapper instance.
+	 */
+	public HeaderMapper<S> getHeaderMapper() {
```
We could change to a configuration-style API here, for example `configureHeaderMapper(Consumer<HeaderMapper<S>> configurer)`.

That preserves encapsulation and keeps the API aligned with the configuration patterns used in the project.

tomazfernandes commented on spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/listener/source/AbstractSqsMessageSource.java:
```diff
@@ -109,6 +114,17 @@ protected void doConfigure(ContainerOptions<?, ?> containerOptions) {
+		configureHeaderMapper(sqsContainerOptions);
+	}
+
+	private void configureHeaderMapper(SqsContainerOptions sqsContainerOptions) {
```
Let's move the configuration logic to `SqsMessageIdResolver` and centralize all message id resolution logic there and keep behavior consistent between the template and message source.

I suggest something along these lines:
```java
 public static void configureMessageIdResolution(
          MessagingMessageConverter<Message> converter, boolean convertMessageIdToUuid)`.
```

tomazfernandes commented on spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/operations/SqsTemplateOptions.java:
```diff
@@ -24,6 +24,7 @@
 /**
  * Sqs specific options for the {@link SqsTemplate}.
  *
+ * @author Jeongmin Kim
```
Please add me as an author here as well, along with `@since 3.0.0`. I missed it when the class was originally created.

tomazfernandes commented on spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/support/converter/SqsMessageIdResolver.java:
```diff
@@ -0,0 +1,98 @@
+		MessageHeaders withRawId = MessageHeaderUtils.addHeaderIfAbsent(headers, SqsHeaders.SQS_RAW_MESSAGE_ID_HEADER,
+				messageId);
+		if (isValidUuid(messageId)) {
+			return new MessagingMessageHeaders(withRawId, UUID.fromString(messageId));
```
Not critical, but we're converting `UUID.fromString` twice.

Perhaps the `isValidUuid` method could return an `Optional<UUID>` so we convert it only once. Method name might need to be adjusted as well.

tomazfernandes commented on spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/operations/SqsTemplate.java:
```diff
@@ -327,16 +334,21 @@ private CompletableFuture<Boolean> handleAutoDeduplication(String endpointName)
+		Map<String, Object> additionalInfo = new HashMap<>();
+		if (sequenceNumber != null) {
+			additionalInfo.put(SqsTemplateParameters.SEQUENCE_NUMBER_PARAMETER_NAME, sequenceNumber);
+		}
```
Looks good, but we're returning the original message which does not contain the new header.

tomazfernandes replied:
@jm0514, this last change is causing `shouldSendAndReceiveBatchFifo` integration test to break.

This raises an important point: while returning the message with the id from the broker would make ids consistent, returning the original message in the `SendResult` allows the sender to correlate the send results with the original `Message` instances they sent, which is a stronger reason for keeping the original behavior.

Let's go with your original direction - add the original message to `SendResult`, and have an additional parameter for the raw message id. Please update the documentation accordingly.

Thanks.

jm0514 replied:
Restored the original message in `SendResult`, and added the raw message ID to `additionalInformation` via `SQS_RAW_MESSAGE_ID_PARAMETER_NAME`. Updated the documentation and tests accordingly.

jm0514 commented:
@tomazfernandes Thanks for catching these — I missed the header propagation issue in SendResult while refactoring.
 
 Here's a summary of the changes:                                                                                                                     
                             
  - `SqsMessageIdResolver`: Renamed `isValidUuid` to `tryParseUuid` returning `Optional<UUID>` to avoid duplicate `UUID.fromString` parsing.                     
  - `SqsTemplate`: `createSendResult` now builds a new message with `resolvedHeaders` so `Sqs_RawMessageId` is included in `SendResult.message()`.
  - `AbstractListenerObservation`: Use `getRawMessageId()` in `Context` so observation tags reflect the broker ID for non-UUID environments.                 
  - `Doc`: Updated to clarify that `Sqs_RawMessageId` header is always present regardless of the `convert-message-id-to-uuid` flag.

tomazfernandes commented:
@maciejwalkowiak, the CI is failing due to Localstack requiring auth: https://github.com/awspring/spring-cloud-aws/actions/runs/23697734758/job/69074091139?pr=1433#step:4:7455

Related: https://github.com/awspring/spring-cloud-aws/issues/1586#issuecomment-4148136287

I've ran the tests locally and they pass, so I'll proceed with the merge:
<img width="1640" height="486" alt="image" src="https://github.com/user-attachments/assets/a46003f2-30be-418a-9a7e-f93d4df6cec5" />

jm0514 commented:
Thanks for the review and merge!
